From ff8436cd1937eabf6e0589e20a134f9c9061a5f3 Mon Sep 17 00:00:00 2001
From: Nicolas Toromanoff <nicolas.toromanoff@foss.st.com>
Date: Wed, 2 Feb 2022 22:04:51 +0100
Subject: [PATCH 0309/1695] plat-stm32mp1: add new API to erase SRAM3

Add new API TEE_Result stm32mp_syscfg_erase_sram3(void);
to be able to erase SRAM3 by hardware request.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git da7864f0393330b9d59c428558e52cb298b0ee0a plat-stm32mp1: add new API to erase SRAM3]
Signed-off-by: Nicolas Toromanoff <nicolas.toromanoff@foss.st.com>
Change-Id: Iaa897f5906192c73233ee86023a48579736ea259
---
 .../plat-stm32mp1/drivers/stm32mp1_syscfg.c   | 46 +++++++++++++++++++
 core/arch/arm/plat-stm32mp1/stm32_util.h      |  3 ++
 2 files changed, 49 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
index edf32c42d..03e02b2b1 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
@@ -18,6 +18,8 @@
 /*
  * SYSCFG register offsets (base relative)
  */
+#define SYSCFG_SRAM3ERASER			U(0x10)
+#define SYSCFG_SRAM3KR				U(0x14)
 #define SYSCFG_IOCTRLSETR			U(0x18)
 #define SYSCFG_CMPCR				U(0x20)
 #define SYSCFG_CMPENSETR			U(0x24)
@@ -36,6 +38,17 @@
 #define SYSCFG_IOCTRLSETR_HSLVEN_SDMMC		BIT(3)
 #define SYSCFG_IOCTRLSETR_HSLVEN_SPI		BIT(4)
 
+/*
+ * SYSCFG_SRAM3ERASE Register
+ */
+#define SYSCFG_SRAM3KR_KEY1			U(0xCA)
+#define SYSCFG_SRAM3KR_KEY2			U(0x53)
+
+#define SYSCFG_SRAM3ERASER_SRAM3EO		BIT(1)
+#define SYSCFG_SRAM3ERASER_SRAM3ER		BIT(0)
+
+#define SYSCFG_SRAM3ERASE_TIMEOUT_US		U(1000)
+
 /*
  * SYSCFG_CMPCR Register
  */
@@ -214,3 +227,36 @@ void stm32mp_enable_fixed_vdd_hslv(void)
 		   SYSCFG_IOCTRLSETR_HSLVEN_SPI);
 }
 #endif
+
+TEE_Result stm32mp_syscfg_erase_sram3(void)
+{
+	vaddr_t base = get_syscfg_base();
+	uint64_t timeout_ref = 0;
+
+	if (!IS_ENABLED(CFG_STM32MP13))
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	/* Unlock SYSCFG_SRAM3ERASER_SRAM3ER */
+	io_write32(base + SYSCFG_SRAM3KR, SYSCFG_SRAM3KR_KEY1);
+	io_write32(base + SYSCFG_SRAM3KR, SYSCFG_SRAM3KR_KEY2);
+
+	/* Request SRAM3 erase */
+	io_setbits32(base + SYSCFG_SRAM3ERASER, SYSCFG_SRAM3ERASER_SRAM3ER);
+
+	/* Lock SYSCFG_SRAM3ERASER_SRAM3ER */
+	io_write32(base + SYSCFG_SRAM3KR, 0);
+
+	/* Wait end of SRAM3 erase */
+	timeout_ref = timeout_init_us(SYSCFG_SRAM3ERASE_TIMEOUT_US);
+	while (io_read32(base + SYSCFG_SRAM3ERASER) &
+	       SYSCFG_SRAM3ERASER_SRAM3EO) {
+		if (timeout_elapsed(timeout_ref))
+			break;
+	}
+
+	/* Timeout may append due to a schedule after the while(test) */
+	if (io_read32(base + SYSCFG_SRAM3ERASER) & SYSCFG_SRAM3ERASER_SRAM3EO)
+		return TEE_ERROR_BUSY;
+
+	return TEE_SUCCESS;
+}
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index d2e30e34f..ac37a26f6 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -33,6 +33,9 @@ vaddr_t stm32mp_bkpreg(unsigned int idx);
 /* Get device ID from SYSCFG registers */
 uint32_t stm32mp_syscfg_get_chip_dev_id(void);
 
+/* Erase ESRAM3 */
+TEE_Result stm32mp_syscfg_erase_sram3(void);
+
 /* Platform util for the GIC */
 vaddr_t get_gicd_base(void);
 
-- 
2.47.2

